### PR TITLE
Implement HGCal "Multiclusters"

### PR DESCRIPTION
**Pull request by lgray**
This PR implements HGCal clusters made of multiple layer clusters. 
These, eventually, will form the basis of physics object reconstruction and analysis using HGCal.
The HGCalMultiCluster class itself is derived from PFCluster so that it can be easily used later in superclustering and particle flow. 

Using an igprof report from #16813 there was a simple optimization to vastly speed up the layer clusters as well.

Attn:@emeschi

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

File: DataFormats/ParticleFlowReco/interface/HGCalMultiCluster.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  double simple_z(double vz) const;    
+  double simple_slope_x(double vz) const; 
+  double simple_slope_y(double vz) const; 
+  
+  double simple_eta(double vz) const;     
+  double simple_phi() const; 
+  double simple_rho() const;
+  
+  double total_uncalibrated_energy() const{
```
Comment (slava77) [inline DataFormats/ParticleFlowReco/interface/HGCalMultiCluster.h]: this is not necessary: constructors set Cluster::energy_ to this value and there is an accessor to it

File: DataFormats/ParticleFlowReco/interface/HGCalMultiCluster.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    double acc = 0.0;
+    for(const auto& ptr : myclusters ) {
+      acc += ptr->energy();
+    }
+    return acc;
+  }
+  bool operator > (const HGCalMultiCluster& rhs) const { 
+    return (total_uncalibrated_energy() > rhs.total_uncalibrated_energy());
```
Comment (slava77) [inline DataFormats/ParticleFlowReco/interface/HGCalMultiCluster.h]: use ```energy()``` no need to invent new interface

**Comment by lgray** [comment]
@cmsbuild please test

File: RecoLocalCalo/HGCalRecProducers/python/hgcalLayerClusters_cfi.py [hunk 3]
```diff
@@ -8,8 +8,13 @@
     detector = cms.string("all"),
     doSharing = cms.bool(False),
     deltac = cms.double(2.),
-    ecut = cms.double(0.06),
+    ecut = cms.double(0.01),
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/python/hgcalLayerClusters_cfi.py]: is this intentional?
I recall  that 0.06 was driven by CPU time.

**Comment by lgray** [comment]
This PR includes a patch that drastically reduces the time spent in the
per-layer clustering.

File: RecoLocalCalo/HGCalRecAlgos/src/ClusterTools.cc [hunk 4]
```diff
@@ -74,3 +74,34 @@ float ClusterTools::getClusterHadronFraction(const reco::CaloCluster& clus) cons
...
+    const double x = ptr->x();    
+    const double y = ptr->y();
+    const float point_r = std::sqrt(x*x + y*y);    
+    const double point_z = ptr->z()-vz;
+    const double weight = ptr->energy() * ptr->size();
+    assert((y != 0. || x != 0.) && "Cluster position somehow in beampipe.");
+    assert(point_z != 0. && "Layer-cluster position given as reference point.");
+    acc_rho += point_r * weight;
+    acc_phi += std::atan2(y,x) * weight;
+    acc_eta += -1. * std::log(std::tan(0.5*std::atan2(point_r,point_z))) * weight;
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecAlgos/src/ClusterTools.cc]: reducing this to acosh or using vdt may be better, to be checked with igprof  if there is much weight here.

**Comment by slava77** [comment]
On the technical side:
- disk size: change in threshold leads to 3-fold increase in size of CaloClusters hgcalLayerClusters, now up to 1.3MB/evt. This should still be manageable for RECO, but by the time we move to AOD, some reduced cluster collection will be needed
- CPU time of hgcalLayerClusters has increase by about 2 from 17 s to 31 s per event (about 2% of production-like RECO job). Somewhat OK.

**Comment by slava77** [comment]
new collections are pretty small in size
```
bytes/evt
343          recoHGCalMultiClusters_hgcalLayerClusters_sharing_RECO.
7838          recoHGCalMultiClusters_hgcalLayerClusters__RECO.

```

**Comment by slava77** [comment]
Some plots:
(baseline is in **red**)

PU200
all_orig-2023pu200vssign809-2023pu200_ttbar14tev2023d4timingpuwf23034p0c_log10recocaloclusters_hgcallayerclusters__reco_obj_energy
Curiously, population with energies above 1 GeV has increased as well

Multicluster distribution looks somewhat healthy
all_orig-2023pu200vssign809-2023pu200_ttbar14tev2023d4timingpuwf23034p0c_log10recohgcalmulticlusters_hgcallayerclusters__reco_obj_energy
all_orig-2023pu200vssign809-2023pu200_ttbar14tev2023d4timingpuwf23034p0c_recohgcalmulticlusters_hgcallayerclusters__reco_obj_eta

the sharing collection  looks empty (size is 0)

====================

QCD600to800 sample
all_orig-histatvssign809-histat_qcd600to800in14tev2023d4wf21253p0c_log10recohgcalmulticlusters_hgcallayerclusters__reco_obj_energy

all_orig-histatvssign809-histat_qcd600to800in14tev2023d4wf21253p0c_recohgcalmulticlusters_hgcallayerclusters__reco_obj_eta

vs layer clusters which appear to be noise-dominated:
all_orig-histatvssign809-histat_qcd600to800in14tev2023d4wf21253p0c_recocaloclusters_hgcallayerclusters__reco_obj_eta

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalClusterTestProducer.cc [hunk 5]
```diff
@@ -19,7 +19,6 @@
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalClusterTestProducer.cc]: Is there a good reason to keep calling this *Test* ? 
HGCalClusterTestProducer

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalClusterTestProducer.cc [hunk 6]
```diff
@@ -152,8 +164,32 @@ void HGCalClusterTestProducer::produce(edm::Event& evt,
...
+  for( unsigned i = 0; i < clusterHandleSharing->size(); ++i ) {
+    edm::Ptr<reco::BasicCluster> ptr(clusterHandleSharing,i);
+    clusterPtrsSharing.push_back(ptr);
+  }
+
+  std::unique_ptr<std::vector<reco::HGCalMultiCluster> > 
+    multiclusters( new std::vector<reco::HGCalMultiCluster> ), 
+    multiclusters_sharing( new std::vector<reco::HGCalMultiCluster> );
+
+  *multiclusters = std::move(multicluster_algo->makePreClusters(clusterPtrs));
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalClusterTestProducer.cc]: remind me please what is the meaning of "Pre" here ?
Is there going to be another clustering algorithm on top of this?

File: RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc [hunk 7]
```diff
@@ -172,15 +172,15 @@ math::XYZPoint HGCalImagingAlgo::calculatePosition(std::vector<KDNode> &v){
 } 
 
 double HGCalImagingAlgo::distance(const Hexel &pt1, const Hexel &pt2){
-  const GlobalPoint position1( std::move( rhtools_.getPosition( pt1.detid ) ) );
-  const GlobalPoint position2( std::move( rhtools_.getPosition( pt2.detid ) ) );
-  return sqrt(pow(pt1.x - pt2.x, 2) + pow(pt1.y - pt2.y, 2));
+  const double dx = pt1.x - pt2.x;
+  const double dy = pt1.y - pt2.y;
+  return std::sqrt(dx*dx + dy*dy);
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc]: is distance needed primarily to select or order elements?
If the actual sqrt is needed in a much smaller fraction of cases, better compute distance^2
and the apply sqrt in that smaller fraction of cases.

**Comment by slava77** [comment]
igprof

https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_9_0_X_2016-12-10-1100-sign809.2023PU200.23034.0_TTbar_14TeV.step3_10_mt1_pp/112

the log/atan mentioned earlier is about 1/3 of the module time


As discussed in skype, these can go to the next PR

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16909 654aa90
- code changes are in line with the description and follow up review; the changes affect only hgcalLayerClusters (not yet used downstream)
- jenkins tests pass and comparisons show no differences
- local tests show some increase in size (expected) and a moderate increase in CPU (see notes posted earlier)

Some technical improvements are left for a later PR.

**Comment by davidlange6** [comment]
+1

**Pull request merged**